File: client/android/src/main/java/com/nexus/app/security/SecurityGuard.kt
package com.nexus.app.security

import android.util.Log
import androidx.annotation.Keep
import java.util.regex.PatternSyntaxException

@Keep
class SecurityGuard private constructor() {

    data class SanitizedInput(
        val original: String,
        val sanitized: String,
        val commandString: String,
        val threatLevel: ThreatLevel,
        val rejected: Boolean,
        val reason: String = ""
    )

    companion object {
        @Volatile
        private var INSTANCE: SecurityGuard? = null

        fun getInstance(): SecurityGuard {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: SecurityGuard().also { INSTANCE = it }
            }
        }
    }

    private val forbiddenTokens = setOf(
        "ignore previous",
        "ignore the above",
        "disregard",
        "override instructions",
        "system prompt",
        "you are now",
        "pretend you are",
        "act as a",
        "jailbreak",
        "dar mode",
        "developer mode",
        "ignore safety",
        "new instruction",
        "leak",
        "reveal prompt",
        "<script",
        "</script>",
        "javascript:",
        "onerror=",
        "onload=",
        "{{",
        "{%",
        "\${",
        "<%",
        "<?php",
        "exec(",
        "eval(",
        "system(",
        "rm -rf",
        "sudo ",
        "chmod ",
        "wget ",
        "curl ",
        "--disable",
        "--no-sandbox"
    )

    private val dangerousPatterns = listOf(
        Regex("\\b(?:ignore|disregard|override).{0,40}(?:instructions?|prompt|command)", RegexOption.IGNORE_CASE),
        Regex("\\b(?:you are|pretend|act as).{0,30}(?:admin|root|developer|ai|assistant)", RegexOption.IGNORE_CASE),
        Regex("<[^>]+(?:on\\w+\\s*=|javascript:)", RegexOption.IGNORE_CASE),
        Regex("\\$\\{[^}]*\\}"),
        Regex("\\b(?:drop|delete|truncate|insert|update|select).{0,30}(?:table|database|from|into)", RegexOption.IGNORE_CASE),
        Regex("(?:\\b(?:https?|ftp)://)[^\\s]+")
    )

    private val maxLength = 500
    private val urlPattern = Regex("(?:https?|ftp)://[^\\s]+")

    fun scan(input: String): SanitizedInput {
        val trimmed = input.trim().take(maxLength)
        if (trimmed.isBlank()) {
            return SanitizedInput(trimmed, trimmed, trimmed, ThreatLevel.SAFE, false)
        }

        val lowered = trimmed.lowercase()

        forbiddenTokens.forEach { token ->
            if (lowered.contains(token.lowercase())) {
                return SanitizedInput(
                    trimmed,
                    "",
                    "",
                    ThreatLevel.DANGEROUS,
                    true,
                    "Forbidden token detected: $token"
                )
            }
        }

        var suspiciousHits = 0
        dangerousPatterns.forEach { pattern ->
            try {
                if (pattern.containsMatchIn(trimmed)) {
                    suspiciousHits++
                }
            } catch (_: PatternSyntaxException) {
                // ignore malformed patterns
            }
        }

        if (trimmed.length > 300) suspiciousHits++
        if (urlPattern.containsMatchIn(trimmed)) suspiciousHits++

        val threatLevel = when {
            suspiciousHits == 0 -> ThreatLevel.SAFE
            suspiciousHits <= 2 -> ThreatLevel.SUSPICIOUS
            else -> ThreatLevel.DANGEROUS
        }

        val sanitized = when (threatLevel) {
            ThreatLevel.SAFE -> removeControlChars(trimmed)
            else -> sanitizeString(trimmed)
        }

        val commandString = removeControlChars(trimmed)

        if (threatLevel == ThreatLevel.DANGEROUS) {
            return SanitizedInput(
                trimmed,
                sanitized,
                "",
                threatLevel,
                true,
                "Input contains dangerous patterns and has been rejected."
            )
        }

        return SanitizedInput(trimmed, sanitized, commandString, threatLevel, false)
    }

    fun isSafe(input: String): Boolean {
        val result = scan(input)
        return result.threatLevel == ThreatLevel.SAFE && !result.rejected
    }

    private fun removeControlChars(input: String): String {
        return input.replace(Regex("[\u0000-\u001F\\x7F]"), "").trim()
    }

    private fun sanitizeString(input: String): String {
        return input
            .replace(Regex("[\u0000-\u001F\\x7F]"), "")
            .replace("<", "&lt;")
            .replace(">", "&gt;")
            .replace("\"", "&quot;")
            .replace("'", "&#x27;")
            .trim()
    }
}

File: client/desktop/security_guard.py
"""Port of Android SecurityGuard for input sanitization."""

from __future__ import annotations

from typing import ClassVar, List, Optional

from .models import SanitizedInput, ThreatLevel


class SecurityGuard:
    _instance: ClassVar[Optional["SecurityGuard"]] = None
    _initialized: bool = False

    def __new__(cls) -> "SecurityGuard":
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._initialized = False
        return cls._instance

    def __init__(self) -> None:
        if self._initialized:
            return
        self._initialized = True
        self._forbidden_tokens = [
            "ignore previous",
            "ignore the above",
            "disregard",
            "override instructions",
            "system prompt",
            "you are now",
            "pretend you are",
            "act as a",
            "jailbreak",
            "dar mode",
            "developer mode",
            "ignore safety",
            "new instruction",
            "leak",
            "reveal prompt",
            "<script",
            "</script>",
            "javascript:",
            "onerror=",
            "onload=",
            "{{",
            "{%",
            "${",
            "<%",
            "<?php",
            "exec(",
            "eval(",
            "system(",
            "rm -rf",
            "sudo ",
            "chmod ",
            "wget ",
            "curl ",
            "--disable",
            "--no-sandbox",
        ]
        self._dangerous_patterns = [
            "\\b(?:ignore|disregard|override).{0,40}(?:instructions?|prompt|command)",
            "\\b(?:you are|pretend|act as).{0,30}(?:admin|root|developer|ai|assistant)",
            "<[^>]+(?:on\\w+\\s*=|javascript:)",
            "\\$\\{[^}]*\\}",
            "\\b(?:drop|delete|truncate|insert|update|select).{0,30}(?:table|database|from|into)",
            "(?:\\b(?:https?|ftp)://)[^\\s]+",
        ]
        self._compiled_patterns = [
            (pattern, __import__("re").compile(pattern, __import__("re").IGNORECASE))
            for pattern in self._dangerous_patterns
        ]
        self._max_length = 500
        self._control_regex = __import__("re").compile("[\u0000-\u001F\\x7F]")
        self._url_pattern = __import__("re").compile("(?:https?|ftp)://[^\\s]+")

    def scan(self, input: str) -> SanitizedInput:
        trimmed = input.strip()[: self._max_length]
        if not trimmed:
            return SanitizedInput(trimmed, trimmed, trimmed, ThreatLevel.SAFE, False)

        lowered = trimmed.lower()
        for token in self._forbidden_tokens:
            if token.lower() in lowered:
                return SanitizedInput(
                    trimmed,
                    "",
                    "",
                    ThreatLevel.DANGEROUS,
                    True,
                    f"Forbidden token detected: {token}",
                )

        suspicious_hits = 0
        for _, compiled in self._compiled_patterns:
            if compiled.search(trimmed):
                suspicious_hits += 1

        if len(trimmed) > 300:
            suspicious_hits += 1
        if self._url_pattern.search(trimmed):
            suspicious_hits += 1

        if suspicious_hits == 0:
            threat = ThreatLevel.SAFE
        elif suspicious_hits <= 2:
            threat = ThreatLevel.SUSPICIOUS
        else:
            threat = ThreatLevel.DANGEROUS

        if threat == ThreatLevel.SAFE:
            sanitized = self._remove_control_chars(trimmed)
        else:
            sanitized = self._sanitize_string(trimmed)

        command_string = self._remove_control_chars(trimmed)

        if threat == ThreatLevel.DANGEROUS:
            return SanitizedInput(
                trimmed,
                sanitized,
                "",
                threat,
                True,
                "Input contains dangerous patterns and has been rejected.",
            )

        return SanitizedInput(trimmed, sanitized, command_string, threat, False)

    def is_safe(self, input: str) -> bool:
        result = self.scan(input)
        return result.threat_level == ThreatLevel.SAFE and not result.rejected

    def _remove_control_chars(self, value: str) -> str:
        return self._control_regex.sub("", value).strip()

    def _sanitize_string(self, value: str) -> str:
        return (
            self._control_regex.sub("", value)
            .replace("<", "&lt;")
            .replace(">", "&gt;")
            .replace('"', "&quot;")
            .replace("'", "&#x27;")
            .strip()
        )
